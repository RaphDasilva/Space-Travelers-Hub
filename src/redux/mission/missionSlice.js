import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const getMission = createAsyncThunk('mission/getMission', async () => {
  const result = await axios('https://api.spacexdata.com/v3/missions');
  return result.data;
});

const initialState = {
  mission: [],
  status: 'idle',
  error: null,
};

const missionSlice = createSlice({
  name: 'mission',
  initialState,
  reducers: {
    joinMission: (state, action) => {
      const updatedMissions = state.mission.map((mission) => {
        if (mission.mission_id === action.payload) {
          return { ...mission, reserved: true };
        }
        return mission;
      });
      return { ...state, mission: updatedMissions };
    },
    leavingMission: (state, action) => {
      const updatedMissions = state.mission.map((mission) => {
        if (mission.mission_id === action.payload) {
          return { ...mission, reserved: false };
        }
        return mission;
      });
      return { ...state, mission: updatedMissions };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMission.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(getMission.fulfilled, (state, action) => ({
        ...state,
        mission: action.payload,
        status: 'succeeded',
      }))
      .addCase(getMission.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }));
  },
});

export const missionActions = missionSlice.actions;
export default missionSlice.reducer;
