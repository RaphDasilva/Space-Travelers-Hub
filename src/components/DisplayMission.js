import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styles from '../styles/DisplayMission.module.css';
import { getMission, missionActions } from '../redux/mission/missionSlice';

const DisplayMission = () => {
  const { mission, status, id } = useSelector((store) => store.mission);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getMission());
  }, [dispatch]);

  let outPut;
  if (status === 'loading') {
    outPut = (
      <div className="loading">
        <h5>Loading....</h5>
      </div>
    );
  }

  const joinMissionHandler = (e) => {
    const { id } = e.target.dataset;
    dispatch(missionActions.joinMission(id));
  };

  return (
    <>
      <div>{outPut}</div>
      {
        mission ? (
          mission.map((item) => (
            <div className={styles.missionItems} key={item.mission_id}>
              <div className={styles.missionHead}>
                <h3>{item.mission_name}</h3>
              </div>
              <div className={styles.missionInfo}>
                <p>
                  {item.description}
                </p>
              </div>
              <div className={styles.buttons}>
                <button type="button" className={styles.memberBtn}>
                  NOT A MEMBER
                </button>
              </div>
              <div className={styles.buttons}>
                <button type="button" data-id={id} className={styles.joinMissionBtn} onClick={joinMissionHandler}>
                  Join Mission
                </button>
              </div>
            </div>
          ))
        ) : (
          <div className="loading">
            <h5>Loading....</h5>
          </div>
        )
      }
    </>
  );
};

export default DisplayMission;
