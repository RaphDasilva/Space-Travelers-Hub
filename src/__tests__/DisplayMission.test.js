import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import DisplayMission from '../components/DisplayMission';
import { missionActions } from '../redux/mission/missionSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('DisplayMission component', () => {
  const dispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
  });

  afterEach(() => {
    useDispatch.mockReset();
    useSelector.mockReset();
    dispatch.mockClear();
  });

  it('should render correctly', () => {
    useSelector.mockReturnValue({
      mission: [
        {
          mission_id: 'mission-1',
          mission_name: 'Mission 1',
          description: 'Mission 1 description',
          reserved: false,
        },
        {
          mission_id: 'mission-2',
          mission_name: 'Mission 2',
          description: 'Mission 2 description',
          reserved: true,
        },
      ],
      status: 'idle',
    });

    const { container } = render(<DisplayMission />);
    expect(container).toMatchSnapshot();
  });

  it('should dispatch joinMission action when Join Mission button is clicked', () => {
    useSelector.mockReturnValue({
      mission: [
        {
          mission_id: 'mission-1',
          mission_name: 'Mission 1',
          description: 'Mission 1 description',
          reserved: false,
        },
      ],
      status: 'idle',
    });

    const { getByText } = render(<DisplayMission />);
    const joinMissionBtn = getByText('Join Mission');

    fireEvent.click(joinMissionBtn);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(missionActions.joinMission('mission-1'));
  });

  it('should dispatch leavingMission action when Leave Mission button is clicked', () => {
    useSelector.mockReturnValue({
      mission: [
        {
          mission_id: 'mission-1',
          mission_name: 'Mission 1',
          description: 'Mission 1 description',
          reserved: true,
        },
      ],
      status: 'idle',
    });

    const { getByText } = render(<DisplayMission />);
    const leaveMissionBtn = getByText('Leave Mission');

    fireEvent.click(leaveMissionBtn);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(missionActions.leavingMission('mission-1'));
  });
});
