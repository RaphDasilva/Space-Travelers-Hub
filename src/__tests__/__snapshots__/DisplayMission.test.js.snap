// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayMission component should render correctly 1`] = `
<div>
  <div />
  <div
    class="missionItems"
  >
    <div
      class="missionHead"
    >
      <h3>
        Mission 1
      </h3>
    </div>
    <div
      class="missionInfo"
    >
      <p>
        Mission 1 description
      </p>
    </div>
    <div
      class="buttons"
    >
      <button
        class="memberBtn"
        type="button"
      >
        Not A Member
      </button>
    </div>
    <div
      class="buttons"
    >
      <button
        class="joinMissionBtn"
        data-id="mission-1"
        type="button"
      >
        Join Mission
      </button>
    </div>
  </div>
  <div
    class="missionItems"
  >
    <div
      class="missionHead"
    >
      <h3>
        Mission 2
      </h3>
    </div>
    <div
      class="missionInfo"
    >
      <p>
        Mission 2 description
      </p>
    </div>
    <div
      class="buttons"
    >
      <button
        class="active"
        type="button"
      >
        Active Member
      </button>
    </div>
    <div
      class="buttons"
    >
      <button
        class="leavingMission"
        data-id="mission-2"
        type="button"
      >
        Leave Mission
      </button>
    </div>
  </div>
</div>
`;
